Remove unnecessary rounding leading to subpixel issues in elementFromPoint hit testing

Ahmad-S792 opened the pull request:
#### 5650bc3b7e350f5aff1a106dd058c82246b8623c
<pre>
Remove unnecessary rounding leading to subpixel issues in elementFromPoint hit testing

<a href="https://bugs.webkit.org/show_bug.cgi?id=284698">https://bugs.webkit.org/show_bug.cgi?id=284698</a>
<a href="https://rdar.apple.com/141496802">rdar://141496802</a>

Reviewed by NOBODY (OOPS!).

This patch removes rounding from hit testing code which leads to sub pixel
issues in elementFromPoint API.

Above fix leads to failure in case of `transformed` elements, so this merges
below Blink fix, which uses `transformed bounding box` for this purpose:

Merge: <a href="https://chromium-review.googlesource.com/c/chromium/src/+/1891110">https://chromium-review.googlesource.com/c/chromium/src/+/1891110</a>

Hit testing uses a 1x1 rect (i.e., &quot;bounding box&quot; )at the hit test
location to test intersection with boxes in the tree. The current
implementation always uses a 1x1 bounding box in the local coordinate
space of the box being tested, which is incorrect -- for example, if the
box is scaled by 100x, then using a 1x1 bounding box in the local space
is equivalent to using a 100x100 bounding box in the viewport, and as a
result we hit the transformed box even if the hit test location is still
100px away from it.

This patch fixes the issue by also transforming the bounding box, so
that it is always equivalent to a 1x1 rect in the viewport.

* Source/WebCore/rendering/HitTestLocation.cpp:
(WebCore::HitTestLocation::HitTestLocation):
(WebCore::HitTestLocation::HitTestLocation): Introduce another helper function
* Source/WebCore/rendering/HitTestLocation.h: Helper function definition
* Source/WebCore/rendering/HitTestingTransformState.cpp:
(HitTestingTransformState::boundsOfMappedQuad):
* Source/WebCore/rendering/HitTestingTransformState.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(hitTestLayerByApplyingTransform):
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/elementFromPoint-subpixel-expected.txt:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5650bc3b7e350f5aff1a106dd058c82246b8623c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108135 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27798 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18218 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113346 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/58625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110098 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28495 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36348 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82102 "Found 1 new test failure: transforms/3d/hit-testing/rotated-hit-test-2.html (failure)") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/58625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111083 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/22578 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/97415 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62534 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/21993 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/15551 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58086 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/91943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/15607 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116471 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35200 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/25917 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91130 "Found 1 new test failure: transforms/3d/hit-testing/rotated-hit-test-2.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35575 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/93692 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90925 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35815 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/13578 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/31004 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35100 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40656 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34834 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38192 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36501 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/HitTestLocation.cpp:
```diff
@@ -28,7 +28,7 @@ HitTestLocation::HitTestLocation() = default;
 
 HitTestLocation::HitTestLocation(const LayoutPoint& point)
     : m_point(point)
-    , m_boundingBox(LayoutRect { flooredIntPoint(point), LayoutSize { 1, 1 } })
+    , m_boundingBox(LayoutRect { point, LayoutSize { 1, 1 } })
```
This is going to produce some interesting results at top/right/bottom/leftPadding() functions where we use rounded point. is that expected? also this class is full of explicit integral snapping calls. While this is good direction, it may break some other content.

Ahmad-S792 replied:
Yes - hence, it is draft and I want to see what breaks. :-)

Ahmad-S792 committed:
Remove unnecessary rounding leading to subpixel issues in elementFromPoint hit testing

https://bugs.webkit.org/show_bug.cgi?id=284698
rdar://141496802

Reviewed by NOBODY (OOPS!).

This patch removes rounding from hit testing code which leads to sub pixel
issues in elementFromPoint API.

Above fix leads to failure in case of `transformed` elements, so this merges
below Blink fix, which uses `transformed bounding box` for this purpose:

Merge: https://chromium-review.googlesource.com/c/chromium/src/+/1891110

Hit testing uses a 1x1 rect (i.e., "bounding box" )at the hit test
location to test intersection with boxes in the tree. The current
implementation always uses a 1x1 bounding box in the local coordinate
space of the box being tested, which is incorrect -- for example, if the
box is scaled by 100x, then using a 1x1 bounding box in the local space
is equivalent to using a 100x100 bounding box in the viewport, and as a
result we hit the transformed box even if the hit test location is still
100px away from it.

This patch fixes the issue by also transforming the bounding box, so
that it is always equivalent to a 1x1 rect in the viewport.

* Source/WebCore/rendering/HitTestLocation.cpp:
(WebCore::HitTestLocation::HitTestLocation):
(WebCore::HitTestLocation::HitTestLocation): Introduce another helper function
* Source/WebCore/rendering/HitTestLocation.h: Helper function definition
* Source/WebCore/rendering/HitTestingTransformState.cpp:
(HitTestingTransformState::boundsOfMappedQuad):
* Source/WebCore/rendering/HitTestingTransformState.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(hitTestLayerByApplyingTransform):
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/elementFromPoint-subpixel-expected.txt:
